DatadogCore/Sources/Core/Upload/BackgroundTaskCoordinator.swift

```diff
@@ -0,0 +1,41 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2019-Present Datadog, Inc.
+ */
+
+import Foundation
+
+internal protocol BackgroundTaskCoordinator {
+    func registerBackgroundTask() -> UUID
+    @discardableResult
+    func endBackgroundTaskIfActive(_ uuid: UUID) -> Bool
+}
+
+#if canImport(UIKit)
+import UIKit
+
+/// The `BackgroundTaskCoordinator` class provides an abstraction for managing background tasks and includes methods for registering and ending background tasks.
+/// It also serves as a useful abstraction for testing purposes.
+internal class UIKitBackgroundTaskCoordinator: BackgroundTaskCoordinator {
+    internal var tasks: [UUID: UIBackgroundTaskIdentifier] = [:]
+
+    internal func registerBackgroundTask() -> UUID {
+        let uuid = UUID()
+        tasks[uuid] = UIApplication.dd.managedShared?.beginBackgroundTask { [weak self] in
+            self?.endBackgroundTaskIfActive(uuid)
+        }
```

Comment:
In [`handler` doc](https://developer.apple.com/documentation/uikit/uiapplication/1623031-beginbackgroundtask) Apple states:
> The system calls the handler synchronously **on the main thread**, blocking the app’s suspension momentarily.

which makes the `internal var tasks:` not thread-safe. When `handler` is invoked on the main thread while `DataUploadWorker` executes `registerBackgroundTask()` (on background thread), this will crash.

This can be solved with a design that doesn't imply shared state, e.g.:
```swift
let backgroundTask = coordinator.register()
backgroundTask.end()
```